### PR TITLE
Validate datastore engine

### PR DESCRIPTION
**Pull request by Himanshu372**
### Added 
1. Added a `DatastoreEngine` type which takes up const permissible engine type values
2. Added helper methods for the type `DatastoreEngine`
3. Added checks to validate engine AFTER it is marshalled into config type
4. Added tests 

## References
Addresses comment - https://github.com/openfga/openfga/pull/712#issuecomment-1574015631
Closes https://github.com/openfga/openfga/issues/681

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [x] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

**Comment by Himanshu372** [comment]
Hi @jon-whit, based on our last discussion, raised this PR to verify Datastore engine. Please have a look. Thanks

File: cmd/shared.go
Comment (jon-whit) [inline cmd/shared.go:1]: Mind renaming to `cmd/cmd.go`?

File: cmd/shared.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+// String converts DatastoreEngine e to string
+func (e DatastoreEngine) String() string {
+	return string(e)
+}
+
+// IsValid checks if DatastoreEngine e is valid
+func (e DatastoreEngine) IsValid() bool {
+	var isValid bool
+	for _, engineType := range []DatastoreEngine{Memory, Postgres, MySQL} {
```
Comment (jon-whit) [inline cmd/shared.go:25]: This code can change to simply:

```go
for _, engine := range []DatastoreEngine{Memory, Postgres, MySQL} {
    if engine.String() == e.String() {
        return true
    }
}

return false
```

The variable assignment and then return is redundant.

File: cmd/shared.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+		if engineType.String() == e.String() {
+			isValid = true
+		}
+	}
+	return isValid
+}
+
+// NewDatastoreEngine inits a new valid DatastoreEngine type
+func NewDatastoreEngine(engine string) (*DatastoreEngine, error) {
```
Comment (jon-whit) [inline cmd/shared.go:34]: Why return a pointer instead of a value type here?

File: cmd/run/run_test.go [hunk 3]
```diff
@@ -283,6 +284,23 @@ func TestVerifyConfig(t *testing.T) {
 		err := VerifyConfig(cfg)
 		require.Error(t, err)
 	})
+
+	t.Run("correct datastore engine specified", func(t *testing.T) {
+		cfg := DefaultConfig()
+		cfg.Datastore.URI = "postgresql://localhost:5432"
```
Comment (jon-whit) [inline cmd/run/run_test.go:290]: I didn't think `postgresql` was a valid prefix? The enumerations in `cmd/shared.go` below define `postgres`.

This test should return an error, and we should change this test to assert `postgres://localhost:5432` does not.

**Review by jon-whit — CHANGES_REQUESTED** [review]
Left some comments for review.

File: cmd/run/run_test.go (hunk 3 above)
Comment (Himanshu372) [reply cmd/run/run_test.go]: I think, I accidently set URI in this test. I just wanted to test for correct DatastoreEngine. After changes, I am getting expected result. 
Although one thing which I noticed is, on setting wrong postgres URI, `RunServer` stalls and waits for connection. I think we should fail early. Let me know your thoughts

**Comment by Himanshu372** [comment]
> Left some comments for review.

@jon-whit Addressed comments, plz check.

**Comment by Himanshu372** [comment]
> > Left some comments for review.
> 
> @jon-whit Addressed comments, plz check.

Hey @jon-whit plz have a look and let me know if the changes are OK

File: cmd/cmd.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+		}
+	}
+	return false
+}
+
+// NewDatastoreEngine inits a new valid DatastoreEngine type
+func NewDatastoreEngine(engine string) (DatastoreEngine, error) {
+	dsEngine := DatastoreEngine(engine)
+	if isValid := dsEngine.IsValid(); !isValid {
+		return "", fmt.Errorf("invalid datastore engine '(%s)'", engine)
```
Comment (miparnisari) [inline cmd/cmd.go:36]: We have similar code here https://github.com/openfga/openfga/blob/0f3e337dec1503fa3fc6b045f2efbb27a4defa19/cmd/run/run.go#L360 and here https://github.com/openfga/openfga/blob/8dc2ed62413e7cccc7da57da057f1f1fc17dcb8a/cmd/validatemodels/validate_models.go#L70

**Comment by miparnisari** [comment]
Hi @Himanshu372, i apologize for taking so long to review this. I took the liberty of updating this PR with respect to main. I also added some more validations to the datastore configs.

**Comment by curfew-marathon** [comment]
Thank you for opening the PR.  At this time we are closing the PR as it has gone stale and needs to be brought up to date.  In the future feel free to re-open the PR.  Thanks again.

**Pull request closed without merging**